File: src/main/java/algorithms/graph/BoundaryOfBinaryTree.java
package algorithms.graph;

import algorithms.TreeNode;

import java.util.LinkedList;
import java.util.List;

/**
 * 545. Boundary of Binary Tree
 */
public class BoundaryOfBinaryTree {


    public List<Integer> boundaryOfBinaryTree(TreeNode root) {
        List<Integer> result = new LinkedList<>();
        result.add(root.val);
        if (root.left == null && root.right == null) {
            return result;
        }
        dfsLeft(root.left, result);
        dfsLeafs(root, result);
        dfsRight(root.right, result);
        return result;
    }

    private void dfsLeft(TreeNode root, List<Integer> result) {
        if (root == null) return;
        if (root.left == null && root.right == null) {
            return;
        }
        result.add(root.val);
        if (root.left != null) {
            dfsLeft(root.left, result);
        } else if (root.right != null) {
            dfsLeft(root.right, result);
        }
    }

    private void dfsLeafs(TreeNode root, List<Integer> result) {
        if (root == null) return;
        if (root.left != null) {
            dfsLeafs(root.left, result);
        }
        if (root.right != null) {
            dfsLeafs(root.right, result);
        }
        if (root.left == null && root.right == null) {
            result.add(root.val);
        }
    }

    private void dfsRight(TreeNode root, List<Integer> result) {
        if (root == null) return;
        if (root.left == null && root.right == null) {
            return;
        }
        if (root.right != null) {
            dfsRight(root.right, result);
        } else if (root.left != null) {
            dfsRight(root.left, result);
        }
        result.add(root.val);
    }

}

File: go/algorithms/boundary_of_binary_tree.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

var leftMostIsFilled = false
var initialRoot *TreeNode

/*
545. Boundary of Binary Tree
*/
func boundaryOfBinaryTree(root *TreeNode) []int {
	array := []int{root.Val}
	if root.Left == nil && root.Right == nil {
		return array
	}
	initialRoot = root
	leftMostIsFilled = false
	dfsBoundary(root, root.Left != nil, root.Right != nil, &array)
	return array
}

func dfsBoundary(root *TreeNode, leftMost bool, rightMost bool, array *[]int) {
	if !leftMostIsFilled && leftMost && (root.Left != nil || root.Right != nil) && initialRoot != root {
		*array = append(*array, root.Val)
	}
	if root.Left != nil && root.Right != nil {
		dfsBoundary(root.Left, leftMost, false, array)
		dfsBoundary(root.Right, false, rightMost, array)
	} else if root.Left != nil {
		dfsBoundary(root.Left, leftMost, rightMost, array)
	} else if root.Right != nil {
		dfsBoundary(root.Right, leftMost, rightMost, array)
	} else {
		leftMostIsFilled = true
		*array = append(*array, root.Val)
	}
	if !leftMost && rightMost && (root.Left != nil || root.Right != nil) && initialRoot != root {
		*array = append(*array, root.Val)
	}
}
